Configuration via environment variables

duijf opened the pull request:
This PR allows us to configure `vaultenv` using environment variables (fixes #28). We support setting the following options:

 - `--host`
 - `--port`
 - `--token`
 - `--secrets-file`
 - `--[no-]connect-tls`
 - `--[no-]validate-certs`
 - `--[no-]inherit-env`
 - `--retry-base-delay-milliseconds`
 - `--retry-attempts`

Notes:

 - The first three are named after the corresponding variables which can be used to configure the official `vault` CLI: `VAULT_HOST`, `VAULT_PORT`, and `VAULT_TOKEN`. 
 - The others are prefixed with `VAULTENV`. 
 - The options which disable defaults are named `VAULTENV_NO_*`, because they explicitly disable the default behavior.
 - CLI arguments take precedence over environment variables.
 - Boolean settings are parsed from the strings `"true"` and `"false"`. No other formats are supported.
 - Because some users might want to override `VAULTENV_NO_*` environment variables, I added in new options: `--validate-certs`, `--inherit-env`, `--connect-tls`, which are no-ops by default, but can override these environment variables. These options are mutually exclusive: vaultenv will throw an error when it encounters both `--connect-tls` and `--no-connect-tls`. (Although the error message isn't that great)

rkrzr commented on app/Main.hs:
```diff
@@ -93,62 +100,143 @@ newtype MilliSeconds = MilliSeconds { unMilliSeconds :: Int }
+--     OptionRecord <$> parser1 <*> parser2 <*> parser3
+-- @
+--
+-- Here, the parser for @OptionRecord@ is the combination of parsers of it's
```
s/it's/its

duijf replied:
Fixed

rkrzr commented on app/Main.hs:
```diff
@@ -93,62 +100,143 @@ newtype MilliSeconds = MilliSeconds { unMilliSeconds :: Int }
-           <> help "maximum number of vault connection retries. Defaults to 9")
+-- | Parser for our CLI options. Seems intimidating, but is straightforward
+-- once you know about applicative parsing patterns. We construct a parser for
+-- @Options@ by concatenating parsers for parts of the record.
```
s/concatenating/combining

duijf replied:
Fixed

rkrzr commented on app/Main.hs:
```diff
@@ -48,13 +49,19 @@ data Options = Options
   , oRetryAttempts   :: Int
   } deriving (Eq, Show)
 
+data EnvSwitches = EnvSwitches
```
Can you add a haddock string for this datatype that explains why we need this?
Esp. why we have both `EnvSwitches` and `EnvVar`.

(we should add haddock comments to everything at some point, but let's at least do so for new code for now)

rkrzr commented on app/Main.hs:
```diff
@@ -93,62 +100,143 @@ newtype MilliSeconds = MilliSeconds { unMilliSeconds :: Int }
+    lookupFromEnvWithDefault key defVal = maybe (value defVal) value (readFromEnvironment key)
+
+    readFromEnvironment :: Read a => String -> Maybe a
+    readFromEnvironment var = lookup var environment >>= Read.readMaybe
```
This parser has become quite massive. I would break it up into smaller parts:

1. Lift the functions in the `where` clause to the top-level. Add a docstring to them.
2. Put each individual parser on the top-level. Also document them (especially the `flag` and `<|>` story).
3. Combine them with `optionsParser envSwitches environment = Options <$> hostParser <*> portParser ...`
4. Maybe move the parsing functions to a different module?

rkrzr commented on app/Main.hs:
```diff
@@ -93,62 +100,143 @@ newtype MilliSeconds = MilliSeconds { unMilliSeconds :: Int }
+  , esNoInheritEnv = envSwitch "VAULTENV_NO_INHERIT_ENV"
+  }
+  where
+    envSwitch key =
```
s/envSwitch/lookupEnvSwitch

rkrzr commented on app/Main.hs:
```diff
@@ -163,25 +251,29 @@ vaultRetryPolicy opts = Retry.fullJitterBackoff (unMilliSeconds (oRetryBaseDelay
+
+  cliAndEnvOptions <- execParser (optionsInfo (parseEnvSwitches localEnvVars) localEnvVars)
+
+  print cliAndEnvOptions
```
Should this still be here?

rkrzr commented on app/Main.hs:
```diff
@@ -163,25 +251,29 @@ vaultRetryPolicy opts = Retry.fullJitterBackoff (unMilliSeconds (oRetryBaseDelay
-  cliOptions <- execParser (optionsInfo localEnvVars)
-  httpManager <- getHttpManager cliOptions
+
+  cliAndEnvOptions <- execParser (optionsInfo (parseEnvSwitches localEnvVars) localEnvVars)
```
I think giving a name to the parser, makes it slightly more readable:

```haskell
let argumentParser = optionsInfo (parseEnvSwitches localEnvVars) localEnvVars
cliAndEnvOptions <- execParser argumentParser
```

duijf commented:
I believe I addressed all your comments. Some notes:

 - I moved the option handling to a separate `Config` module in `src` and added a library section to `package.yaml`
 - I renamed `EnvSwitches` to `EnvFlags` and documented them. The reason for the `EnvVar` type alias was just readability.
 - I didn't put the individual parsers to the toplevel. I opted to put them in the `where` clause instead. Otherwise I'd have to continue to pass in `envFlags` and `envVars` to most subparsers. This way, they are still part of the scope of `optionsParser`

rkrzr commented on app/Main.hs:
```diff
@@ -2,12 +2,16 @@
 
+--
+-- Stackage and stdlib
+--
```
Not really necessary imo, but I also don't really mind.

duijf replied:
Removed

duijf commented on src/Config.hs:
```diff
@@ -112,59 +112,67 @@ optionsParserWithInfo envFlags localEnvVars =
+    <*> secretsFile
+    <*> cmd
+    <*> cmdArgs
+    <*> noConnectTls
```
Sure, will do!

duijf replied:
Done

rkrzr commented on package.yaml:
```diff
@@ -19,10 +26,11 @@ executables:
       - unix
       - utf8-string
-      - mtl
+      - vaultenv
```
I would move all the dependencies here up to the new top-level dependencies section. And only keep `vaultenv` here. It's slightly nicer and also removes duplicates.

duijf replied:
Good suggestion, done!

rkrzr commented on src/Config.hs:
```diff
@@ -0,0 +1,223 @@
+  , oRetryAttempts   :: Int
+  } deriving (Eq, Show)
+
+-- | Behaviour flags that we allow users to set via environment variables.
```
s/Behaviour/Behavior (to be consistent with yourself further down)

duijf replied:
Fixed

rkrzr commented on src/Config.hs:
```diff
@@ -0,0 +1,223 @@
+      parser = optionsParserWithInfo envFlags envVars
+  in OptParse.execParser parser
+
+-- | Parses behaviour flags from a list of environment variables. If an
```
s/behaviour/behavior

rkrzr commented on src/Config.hs:
```diff
@@ -0,0 +1,223 @@
+optionsParserWithInfo :: EnvFlags -> [EnvVar] -> OptParse.ParserInfo Options
+optionsParserWithInfo envFlags localEnvVars =
+  OptParse.info
+    (optionsParser envFlags localEnvVars <**> OptParse.helper)
```
I think `OptParse.helper <*> optionsParser envFlags localEnvVars` should probably work as well here, and  `<*>` is better known than `<**>`.

duijf replied:
Agreed & changed. I think this was an artifact of me following an optparse tutorial way in the beginning

rkrzr commented on src/Config.hs:
```diff
@@ -0,0 +1,223 @@
+      (  metavar "ARGS..."
+      <> help "Arguments to pass to CMD, defaults to nothing")
+    noConnectTls =
+       flag (efNoConnectTls envFlags) True
```
Could you still add a comment to explain the flag logic here? Where does it make sure that command-line flags override env vars?

duijf replied:
I've added some more top level documentation to this function. Not sure if that's the right place, but it is a lot of text and didn't want it inline. Is this better?

rkrzr commented on src/Config.hs:
```diff
@@ -0,0 +1,223 @@
+      <> readValueFromEnvWithDefault "VAULTENV_RETRY_ATTEMPTS" 9 envVars
+      <> help "Maximum number of vault connection retries. Defaults to 9"
+
+-- ! Attempt to parse an optparse default value modifier from a list of
```
s/-- !/-- |  (everywhere in this module)

duijf replied:
Fixed

duijf commented:
Thanks for the reviews! I've taken your suggestions and the docs should be more clear now. Let me know if there's anything else!

rkrzr commented on src/Config.hs:
```diff
@@ -0,0 +1,251 @@
+import Data.Monoid ((<>))
+
+import Options.Applicative (value, long, auto, option, metavar, help, flag,
+                            str, argument, many, strOption, (<**>))
```
The `(<**>)` import can now be removed.

duijf commented:
Oh yeah, I didn't catch that because the `-Wall` and `-Werror` flags were set for the executable only. Fixed that in the latest commit
